### PR TITLE
Add azure blob support

### PR DESCRIPTION
**Pull request by MichaelSlatteryDev**

### CODE REVIEW

File: src/modules/profile/cqrs/command/upload-profile-image.command.ts [hunk 1]
```diff
@@ -13,14 +13,16 @@ export interface UploadProfileImageCommandArgs {
...
         private readonly profileRepository: IProfileRepository,
     ) {}
 
     // TODO: Implement validations and errors
     async execute(args: UploadProfileImageCommandArgs): Promise<void> {
-        const avatar = await this.s3ClientService.upload(args.file, args.id, UploadImageTypeEnum.PROFILE);
+        const avatar = await this.blobClientService.upload(args.file, args.id, UploadImageTypeEnum.PROFILE);
         const profile = await this.profileRepository.findOne({ userId: args.id });
+        // TODO: Not sure how to get an id here? Previously with S3 we could just get the Key but I don't think a similar value
+        // exists in BlockBlobUploadResponse? I just return the file path now
```
Comment (Xazin) [inline src/modules/profile/cqrs/command/upload-profile-image.command.ts:25]: Should be fine, the returned `Key` from S3 was basically the relative location, similar to the one you return now.

File: src/modules/profile/profile.module.ts [hunk 2]
```diff
@@ -8,10 +8,11 @@ import {
     UploadProfileImageCommand,
 } from '@modules/profile/cqrs';
 import { S3Module } from '@modules/core/s3';
+import { BlobModule } from '@modules/core';
 import { ProfileService } from './service';
 
 @Module({
-    imports: [InfrastructureModule, S3Module],
+    imports: [InfrastructureModule, S3Module, BlobModule],
```
Comment (Xazin) [inline src/modules/profile/profile.module.ts:15]: I assume we can completely remove S3Module now

File: src/modules/core/blob/blob-client.service.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        const extension = mime.getExtension(file.mimetype);
+        // TODO: Should all these params still be passed in like the previous implementation?
+        const uploadParams = {
+            Container: this.configService.get('AZURE_CONTAINER'),
+            Key: directory + id + '.' + extension,
+            ContentType: file.mimetype,
+            Body: file.buffer,
+            ACL: 'public-read',
+            ETag: id,
+        };
```
Comment (Xazin) [inline src/modules/core/blob/blob-client.service.ts:30]: Most likely not, see https://learn.microsoft.com/en-us/javascript/api/@azure/storage-blob/blockblobclient?view=azure-node-latest#@azure-storage-blob-blockblobclient-upload

File: src/modules/core/blob/blob-client.repository.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        const azureConnection = this.configService.get('AZURE_CONNECTION');
+        const azureContainer = this.configService.get('AZURE_CONTAINER');
+        const blobServiceClient = BlobServiceClient.fromConnectionString(azureConnection);
+        this.containerClient = blobServiceClient.getContainerClient(azureContainer);
+    }
+
+    async upload(params: any, imageName: string): Promise<BlockBlobUploadResponse> {
+        // TODO: I pass id as the imageName. I presume that should be ok for the name of the file
+        const blobClient = this.containerClient.getBlockBlobClient(imageName);
+        return await blobClient.upload(params, params.length);
```
[the remark this answers is not in the document]
Comment (MichaelSlatteryDev) [reply src/modules/core/blob/blob-client.repository.ts:19]: I changed it to just send file buffer. After testing it out the images are being uploaded to azure so it seems to be good

**Review by Xazin — APPROVED** [review]

**Pull request merged**